proposals/NNNN-keypath-function-subtyping.md

```diff
@@ -0,0 +1,95 @@
+* Review Manager: [John McCall](https://github.com/rjmccall)
+* Status: **Awaiting review**
+* Implementation: [apple/swift#39612](https://github.com/apple/swift/pull/39612)
+* Review: ([pitch](https://forums.swift.org/t/pitch-generalize-keypath-to-function-conversions/52681))
+
+## Introduction
+
+Today, keypath literals can only be narrowly converted to a function which exactly matches the argument and return type. This proposal allows key path literals to partake in the full generality of the conversions we allow between arbitrary function types, so that the following code compiles without error:
+
+```swift
+let _: (String) -> Int? = \.count
+```
+
+## Motivation
+
+[SE-0249](https://github.com/apple/swift-evolution/blob/main/proposals/0249-key-path-literal-function-expressions.md) introduced a conversion between key path literals and function types, which allowed users to write code like the following:
+
+```swift
+let strings = ["Hello", "world", "!"]
+let counts = strings.map(\.count) // [5, 5, 1]
+```
+
+However, SE-0249 does not quite live up to its promise of allowing the equivalent key path construction "wherever it allows (Root) -> Value functions." Function types permit conversions that are covariant in the result type and contravariant in the parameter types, but key path literals require exact type matches. This can lead to some potentially confusing behavior from the compiler:
+
+```swift
+struct S {
+  var x: Int
+}
+
+// All of the following are okay...
+let f1: (S) -> Int = \.x
+let f2: (S) -> Int? = f1
+let f3: (S) -> Int? = { $0.x }
+let f4: (S) -> Int? = { kp in { root in root[keyPath: kp] } }(\S.x)
+let f5: (S) -> Int? = \.x as (S) -> Int
+
+// But the direct conversion fails!
+let f6: (S) -> Int? = \.x // <------------------- Error!
+```
+
+## Proposed solution
+
+Allow key path literals to be converted freely in the same manner as functions are converted today. This would allow the definition `f6` above to compile without error, in addition to allowing constructions like:
+
+```swift
+class Base {
+  var derived: Derived { Derived() }
+}
+class Derived: Base {}
+
+let g1: (Derived) -> Base = \Base.derived
+```
+
+## Detailed design
+
+Rather than permitting a key path literal with root type `Root` and value type `Value` to only be converted to a function type `(Root) -> Value`, key path literals will be permitted to be converted to any function type which `(Root) -> Value` may be converted to.
+
+The actual key-path-to-function conversion transformation proceeds exactly as before, generating code with the following semantics (adapting an example from SE-0249):
+
+```swift
+// You write this:
+let f: (User) -> String? = \User.email
+
+// The compiler generates something like this:
+let f: (User) -> String? = { kp in { root in root[keyPath: kp] } }(\User.email)
+```
+
+## Source compatibility
+
+This proposal makes previously invalid conversions valid, and should not affect source compatibility. In situations such as:
```

Comment:
Unfortunately, these two phrases are contradictory.  Making a previously invalid conversion valid *always* has the potential to affect source compatibility simply because it does make an overload newly viable.  You're right that any previously-viable overload should be preferable *along this dimension*, but overload resolution has to consider multiple dimensions at once, and it's possible that the newly-viable overload is preferable among one or more or those.  As a result, this can introduce ambiguities or even change the interpretation of code.

I think this should be a fairly small problem in practice, but I think you do need to actually do the investigation (in parallel with the review might be okay?), and you *definitely* shouldn't claim in the proposal text that it has no source compatibility implications.

Reply:
How about this:

```
This proposal allows conversions in some situations that were previously impossible.  This can affect source compatibility because overloaded function calls may gain new viable overload candidates.

In typical scenarios, these new candidates will be strictly worse than previous candidates because the new conversion is strictly less favorable.  For example, consider this code:

<your example>

Without this proposal, Swift will prefer to call the generic function because the conversion necessary for the concrete function is invalid.  With this proposal, Swift will still prefer to call the generic function because the concrete function requires an extra conversion (from `(String) -> Bool` to `(String) -> Bool?`).

However, this is not always true. A newly-viable overload candidate may be disfavored for the key path argument but favored for other reasons.  This should be uncommon, and so we expect this proposal will have a very small impact in practice, but this will need to be demonstrated as part of landing the proposal in a Swift release.
```

Reply:
Yeah, you're totally right, this was too aspirational. Updated!